pkg/types/chain_reader.go

```diff
@@ -0,0 +1,85 @@
+package types
+
+import (
+	"context"
+	"time"
+)
+
+type ErrorChainReaderUnsupported struct{}
```

Comment:
I think we can simplify these. I thought it made sense to have them wrap their cause, but now I see the switches to decode over the wire. If we just use a string backed type instead of a struct, then we won't have to do the hacky Error() calls to get the strings to compare.

Alternatively, are we sure there aren't built in GRPC helpers for exactly this use case?

Reply:
I'm confused... where are we comparing strings? I don't recall having done that anywhere.

Reply:
Also, this is not an error that goes across the wire... it's only used internally to the relay.

This is the only place where it's used... if there is another PR somewhere that uses it for some other purpose, I haven't seen it and it shouldn't be considered as part of this PR:
https://github.com/smartcontractkit/chainlink/pull/10990/files#diff-e6ef05d3b1ec007c275a6f1d690c96777f1b58ec80d080a53d63d694aed77e98R532

Reply:
Oh, I see what happened:  I added some types starting with `Error` and then @nolag added some types ending with `Error`.  They're right next to each other in `types/chain_reader.go` but mine are just used internally by the relays while his go over the wire.  So you're probably referring to his types, but seems like we should separate these out more and maybe add comments so it's less confusing:
```
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700  8) type ErrorChainReaderUnsupported struct{}
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700  9) 
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 10) func (e ErrorChainReaderUnsupported) Error() string {
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 11)  return "ChainReader is not supported by the relay"
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 12) }
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 13) 
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 14) type ErrorNoChainReaderInJobSpec struct{}
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 15) 
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 16) func (e ErrorNoChainReaderInJobSpec) Error() string {
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 17)  return "There is no ChainReader configuration defined in the job spec"
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 18) }
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 19) 
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 20) type ErrorChainReaderInvalidConfig struct{}
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 21) 
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 22) func (e ErrorChainReaderInvalidConfig) Error() string {
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 23)  return "Invalid ChainReader configuration"
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 24) }
a9eb8a6d (Domino Valdano 2023-10-27 13:45:45 -0700 25) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 26) type InvalidEncodingError struct{}
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 27) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 28) func (InvalidEncodingError) Error() string {
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 29)  return "cannot decode bytes"
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 30) }
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 31) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 32) type InvalidTypeError struct{}
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 33) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 34) func (InvalidTypeError) Error() string {
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 35)  return "invalid type"
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 36) }
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 37) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 38) type FieldNotFoundError struct{}
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 39) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 40) func (FieldNotFoundError) Error() string {
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 41)  return "field not found"
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 42) }
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 43) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 44) type WrongNumberOfElements struct{}
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 45) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 46) func (WrongNumberOfElements) Error() string {
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 47)  return "wrong number of elements decoded"
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 48) }
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 49) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 50) type NotASliceError struct{}
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 51) 
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 52) func (NotASliceError) Error() string {
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 53)  return "input is not a slice or array"
f0867d8c (Ryan Tinianov  2023-11-02 15:29:13 -0400 54) }
```

Also, is there an established convention we have for whether `Error` should go at the beginning or the end?

Reply:
I typically use a prefix since it keeps them alpha sorted together.